nemo/core/classes/exportable.py

```diff
@@ -0,0 +1,143 @@
+from nemo.core.classes import typecheck
+from nemo.core.neural_types import AxisKind, NeuralType
+
+__all__ = ['ExportFormat', 'Exportable']
+
+
+class ExportFormat(Enum):
+    """Which format to use when exporting a Neural Module for deployment"""
+
+    ONNX = 0
+
+
+class Exportable(ABC):
+    """
+    This Interface should be implemented by particular classes derived from nemo.core.NeuralModule or nemo.core.ModelPT.
+    It gives these entities ability to be exported for deployment to formats such as ONNX.
+    """
+
+    def export(
+        self,
+        output: str,
+        input_example=None,
+        output_example=None,
+        format: ExportFormat = ExportFormat.ONNX,
+        onnx_opset_version=11,
+    ):
+        try:
+            # Disable typechecks
+            typecheck.set_typecheck_enabled(enabled=False)
+
+            _in_example, _out_example = self.prepare_for_export()
+            if input_example is not None:
+                _in_example = input_example
+            if output_example is not None:
+                _out_example = output_example
+
+            # Check if output already exists
+            if os.path.exists(output):
+                raise FileExistsError(f"Destination {output} already exists. " f"Aborting export.")
+
+            if not (hasattr(self, 'input_types') and hasattr(self, 'output_types')):
+                raise NotImplementedError('For export to work you must define input and output types')
+            input_names = list(self.input_types.keys())
+            output_names = list(self.output_types.keys())
+            dynamic_axes = defaultdict(list)
+
+            # extract dynamic axes and remove unnecessary inputs/outputs
+            # for input_ports
+            for _name, ntype in self.input_types.items():
+                if _name in self.disabled_deployment_input_names:
+                    input_names.remove(_name)
+                    continue
+                self.__extract_dynamic_axes(_name, ntype, dynamic_axes)
+            # for output_ports
+            for _name, ntype in self.output_types.items():
+                if _name in self.disabled_deployment_output_names:
+                    output_names.remove(_name)
+                    continue
+                self.__extract_dynamic_axes(_name, ntype, dynamic_axes)
+
+            if len(dynamic_axes) == 0:
+                dynamic_axes = None
+
+            # Set module to eval mode
+            self.eval()
+
+            # Attempt export
+            if format == ExportFormat.ONNX:
+                if _in_example is None:
+                    raise ValueError(f'Example input is None, but ONNX tracing was attempted')
+                if _out_example is None:
+                    if isinstance(_in_example, tuple):
+                        _out_example = self.forward(*_in_example)
+                    else:
+                        _out_example = self.forward(_in_example)
+                with torch.jit.optimized_execution(True):
```

Comment:
That, perhaps, should be optional, if at all.
I think currently the only benefit of this extra step is somewhat better diagnostic. Also, that would be 'whole step' for TorchScript export format. 
What I think makes sense here is adding 'script=False' argument that would attempt jit.script() first (and resort to tracing in TS case or onnx.export in ONNX case). 
That I can actually do later - no need to address in this PR but we should agree on policy.

Reply:
are there any downsides of doing this?

Reply:
I probably misplaced the comment. It was about doing jit.trace() before onnx.export. I would rather do direct onnx export for clarity for now, and then add TS option and try_script argument.